packages/definitions-parser/src/packages.ts

```diff
@@ -83,71 +107,109 @@ export class AllPackages {
   }
 
   /** Gets the latest version of a package. E.g. getLatest(node v6) was node v10 (before node v11 came out). */
-  getLatest(pkg: TypingsData): TypingsData {
+  async getLatest(pkg: TypingsData): Promise<TypingsData> {
     return pkg.isLatest ? pkg : this.getLatestVersion(pkg.typesDirectoryName);
   }
 
-  private getLatestVersion(typesDirectoryName: string): TypingsData {
-    const latest = this.tryGetLatestVersion(typesDirectoryName);
+  private async getLatestVersion(typesDirectoryName: string): Promise<TypingsData> {
+    const latest = await this.tryGetLatestVersion(typesDirectoryName);
     if (!latest) {
       throw new Error(`No such package ${typesDirectoryName}.`);
     }
     return latest;
   }
 
-  tryGetLatestVersion(typesDirectoryName: string): TypingsData | undefined {
-    const versions = this.data.get(typesDirectoryName);
+  async tryGetLatestVersion(typesDirectoryName: string): Promise<TypingsData | undefined> {
+    const versions = await this.tryGetTypingsVersions(typesDirectoryName);
     return versions && versions.getLatest();
   }
 
-  getTypingsData(id: PackageId): TypingsData {
-    const pkg = this.tryGetTypingsData(id);
+  async getTypingsData(id: PackageId): Promise<TypingsData> {
+    const pkg = await this.tryGetTypingsData(id);
     if (!pkg) {
       throw new Error(`No typings available for ${JSON.stringify(id)}`);
     }
     return pkg;
   }
 
-  tryGetTypingsData(pkg: PackageId): TypingsData | undefined {
+  async tryGetTypingsData(pkg: PackageId): Promise<TypingsData | undefined> {
     const typesDirectoryName = pkg.typesDirectoryName ?? pkg.name.slice(scopeName.length + 2);
-    const versions = this.data.get(typesDirectoryName);
+    const versions = await this.tryGetTypingsVersions(typesDirectoryName);
     return (
       versions && versions.tryGet(new semver.Range(pkg.version === "*" ? "*" : `^${formatTypingVersion(pkg.version)}`))
     );
   }
 
-  allPackages(): readonly AnyPackage[] {
-    return [...this.allTypings(), ...this.allNotNeeded()];
+  async tryGetTypingsVersions(typesDirectoryName: string): Promise<TypingsVersions | undefined> {
+    let versions = this.types.get(typesDirectoryName);
+    if (versions) {
+      return versions;
+    }
+    if (this.errors.has(typesDirectoryName)) {
+      return undefined;
+    }
+    const raw = await getTypingInfo(typesDirectoryName, this.dt);
+    if (!raw) {
+      return undefined;
+    }
+    if ("errors" in raw) {
+      this.errors.set(typesDirectoryName, raw.errors);
```

Comment:
this error handling pattern makes me want `tryGetTypingsVersion` and `errors` both to be private. Just in case.

Reply:
I initially had `errors` private... I think I used it externally in some validation that ended up being redundant and I removed it. Will take a look.